### DISCUSSION TITLE
Handling ua.UaError exceptions

### DISCUSSION THREAD

**Discussion by TheFern2** [FreeOpcUa/opcua-asyncio#1059] [Q&A] [answered]
I was going to submit a bug, but wasn't sure if it is. It looks like a bug to me, but maybe not others.

If I understand correctly ua.UaError is the base class for all errors/exceptions, one problem I noted is that if you use ua.UaError you'll get a linter error saying code doesn't exist and rightly so since the base class is empty.

```python
class UaError(RuntimeError):
    pass

class ServiceError(UaError):
    def __init__(self, code):
        super().__init__('UA Service Error')
        self.code = code
```

Shouldn't UaError base class have a dunder init with code, and subclasses use super init? Or is this design deliberate for some other reason?

```python
class UaError(RuntimeError):
    def __init__(self, code):
        self.code = code

class ServiceError(UaError):
    def __init__(self, code):
        super().__init__(code)
```

**Comment by schroeder-** [comment, marked as answer]
It is a error from the linter. When you don't specify a __init__ method, the __init__ from the parent class will be called. In this case the __init__ from RuntimeError is called. RuntimeError __init__ accepts a str as arguments.

**Reply by TheFern2** [reply]
I disagree, I don't think is a linter error, linter is doing exactly what is supposed to, reads UaError class, which is empty and finds no code variable. See usage below. If ua.UaError is the base class for exceptions and has no code, where should the linter look up code?

```python
try:
    value = await opc_node.read_value()
    client_node.value = value
except ua.UaError as e:
    client_node.value = get_name_and_doc(e.code)[0]  # <- code does not exist
    client_node.error = True
```

**Reply by schroeder-** [reply]
Ok I get the problem, but we have errors without code like:

```
class NotEnoughData(UaError):
    pass

class SocketClosedException(UaError):
   pass
```

So there comes ua.UaStatusCodeError into play:
```Python
try:
    value = await opc_node.read_value()
    client_node.value = value
except ua.UaStatusCodeError as e:
    client_node.value = get_name_and_doc(e.code)[0]  # <- code does not exist
    client_node.error = True
except ua.UaError as e:
    client_node.error = True
    client_node.value = ua.StatusCode.BadInternalError
```

**Reply by TheFern2** [reply]
Ah okay I see, I was thinking all errors had a code, and uaError was the base class for all the errors. That makes sense now. Thanks for the details.

**Comment by oroulet** [comment]
These errors are the errors defined by spec. If you can show that the spec says we should raise an UaError in some situations then we should allow it. 
I suppose it does not but i have not checked

**Reply by TheFern2** [reply]
I think @schroeder- answer my question on the other thread. Thanks though.
